[lldb][Formatter] Consolidate libstdc++ and libc++ unique_ptr formatter tests into generic test

Michael137 opened the pull request:
The libc++ test was a subset of the tests in libstdc++. This test moves the libc++ test into `generic` and somne additional test-cases from `libstdc++` (specifically the recursive unique_ptr case). It turns out the libstdc++ formatter supports dereferencing using the "object" or "obj" names. We could either drop those from the tests or support the same for libc++. I took the latter approach but don't have strong opinions on this.

Split out from https://github.com/llvm/llvm-project/pull/146740

Michael137 committed:
[lldb][Formatter] Consolidate libstdc++ and libc++ unique_ptr formatter tests into generic test

The libc++ test was a subset of the tests in libstdc++. This test moves
the libc++ test into `generic` and somne additional test-cases from `libstdc++`
(specifically the recursive unique_ptr case). It turns out the libstdc++
formatter supports dereferencing using the "object" or "obj" names. We
could either drop those from the tests or support the same for libc++. I
took the latter approach but don't have strong opinions on this.

Split out from https://github.com/llvm/llvm-project/pull/146740

labath reviewed (commented):
Looks good modulo the failure in the libstdc++ version. The type name comes out different, presumably because libstdc++ doesn't contain the magic to print the template as std::string.

Testing string name printing is not really relevant for the unique_ptr formatter, so I'd delete the check from there. Testing that `SomeTemplate<std::string>` prints the right way is useful, but would be better of in the test for preferred name printing (if we don't do that already).

Michael137 commented:
> Looks good modulo the failure in the libstdc++ version. The type name comes out different, presumably because libstdc++ doesn't contain the magic to print the template as std::string.
> 
> Testing string name printing is not really relevant for the unique_ptr formatter, so I'd delete the check from there. Testing that `SomeTemplate<std::string>` prints the right way is useful, but would be better of in the test for preferred name printing (if we don't do that already).

Agreed. Removed the typename checks

One final difference between the formatters was that the libstdc++ formatter created a synthetic child for the dereferenced object and calls it `object`. I removed that and made it do what libc++ does: just dereference and present that ValueObject (instead of cloning and renaming it).

Michael137 commented on lldb/test/API/functionalities/data-formatter/data-formatter-stl/generic/unique_ptr/Makefile:
```diff
@@ -1,7 +1,5 @@
-
 # We need debug info tuning for lldb in order to emit the preferred name for
 # std::string. See https://reviews.llvm.org/D145803.
 CXXFLAGS_EXTRAS := -std=c++14 -glldb
```
We can probably drop these too now since we don't check the typenames.

Michael137 committed:
fixup! remove lldb tuning

Michael137 committed:
fixup! loosen test assertions for type-names

Michael137 committed:
fixup! make libstdc++ dereferenced pointer name consistent with libc++

labath commented:
> It turns out the libstdc++ formatter supports dereferencing using the "object" or "obj" names.

I don't think there's any harm in supporting those, though I'd be surprised if many people are aware of that functionality. I kind of like this "hidden child" concept, but it feels like it ought to be just slightly more discoverable.

labath commented:
(The test is failing due to std::make_unique not being defined. I'd be fine both with bringing back -std=c++14 and using explicit construction.)

Michael137 committed:
fixup! compile with c++14
